# German eBay Sales Dataset - Cleaning & Analysis Project 

This project is based on a sample of this [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data) Dataset.

The dataset is first read into pandas using the "Windows-1252" encoding

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding = "Windows-1252")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Exploratory Analysis

The dataset contains 50,000 rows and 20 columns. The columns include information on;

 - dateCrawled - When this ad was first crawled. All field-values are taken from this date.
 - name - Name of the car.
 - seller - Whether the seller is private or a dealer.
 - offerType - The type of listing
 - price - The price on the ad to sell the car.
 - abtest - Whether the listing is included in an A/B test.
 - vehicleType - The vehicle Type.
 - yearOfRegistration - The year in which the car was first registered.
 - gearbox - The transmission type.
 - powerPS - The power of the car in PS.
 - model - The car model name.
 - kilometer - How many kilometers the car has driven.
 - monthOfRegistration - The month in which the car was first registered.
 - fuelType - What type of fuel the car uses.
 - brand - The brand of the car.
 - notRepairedDamage - If the car has a damage which is not yet repaired.
 - dateCreated - The date on which the eBay listing was created.
 - nrOfPictures - The number of pictures in the ad.
 - postalCode - The postal code for the location of the vehicle.
 - lastSeenOnline - When the crawler saw this ad last online.

A look at the initial analysis shows some columns with null entries.

    vehicleType            44905 non-null object
    gearbox                47320 non-null object
    model                  47242 non-null object
    fuelType               45518 non-null object
    notRepairedDamage      40171 non-null object
    
Also, the price and odometer columns are currently the wrong type for further analysis, so cleaning for non-numerical characters and a change of type is a necessary process.

In [4]:
print(autos.columns)
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_pictures', 'postal_code',
       'last_seen']

autos.head()

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  no_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

For easier reading, the format of the column labels was changed. 

### Further Exploratory Analysis

In [5]:
autos.describe(include = "all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  no_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

A number of the columns are sporting only two unique vales showing binary options

In two columns, seller and offer_type, a single value has a frequency of 49,999 out of 50,000. This should be dropped along with the no_pictures column which is has only one value entered.

In [6]:
autos = autos.drop(["no_pictures", "seller", "offer_type"], axis=1)

    Next we look at the price and odometer columns

In [7]:
autos["price"].value_counts().head()

$0        1421
$500       781
$1,500     734
$2,500     643
$1,200     639
Name: price, dtype: int64

In [8]:
autos["odometer"].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In order to explore these columns, there is a need to remove non-numeric characters

In [9]:
autos["price"] = autos["price"].str.replace('[\$,]', '').astype(int)
autos["odometer"] = autos["odometer"].str.replace('[\km,]','').astype(int)

In [10]:
autos.rename(columns = {"odometer":"odometer_km"}, inplace = True)

In [11]:
autos["price"].value_counts().sort_index(ascending = False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             6
35             1
30             7
29             1
25             5
20             4
18             1
17             3
15             2
14             1
13             2
12             3
11             2
10             7
9              1
8              1
5              2
3              1
2             

In [12]:
autos["price"].describe().astype(int)

count       50000
mean         9840
std        481104
min             0
25%          1100
50%          2950
75%          7200
max      99999999
Name: price, dtype: int64

Looking at the price column, there are entries that are unreal in the range of products we are inspecting [0 and variants of 999999 and other numbers]

As such we can safely assume values from 1 to 350000 to be a more valid range considering the auction format of the website.

In [13]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [14]:
autos["odometer_km"].value_counts().sort_index(ascending = False)

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
20000       762
10000       253
5000        836
Name: odometer_km, dtype: int64

In the odometer_km column, there is a reasonable range of values for a car's odometer. With 150,000km, approximately 94,000 miles, being an appropriate number for a particularly old used car and 5,000km is a possible value for a car that is relatively new. 

### Categorical Columns(Translation)

Most of our categorical data is in German, we can explore this values and see if a translation can occur

In [15]:
autos["vehicle_type"].value_counts(dropna=False)

limousine     12598
kleinwagen    10585
kombi          8931
NaN            4586
bus            4031
cabrio         3016
coupe          2463
suv            1965
andere          390
Name: vehicle_type, dtype: int64

In [16]:
autos["gearbox"].value_counts(dropna=False)

manuell      36102
automatik    10120
NaN           2343
Name: gearbox, dtype: int64

In [17]:
autos["fuel_type"].value_counts(dropna=False)

benzin     29368
diesel     14353
NaN         4030
lpg          668
cng           73
hybrid        37
elektro       19
andere        17
Name: fuel_type, dtype: int64

In [18]:
autos["unrepaired_damage"].value_counts(dropna=False)

nein    34775
NaN      9101
ja       4689
Name: unrepaired_damage, dtype: int64

Looking at the columns, we can see that translations are in order. Care should be given to the NaN values as each column has it.

In [19]:
autos["unrepaired_damage"].map({"nein":"no", "ja":"yes", "no":"no", "yes":"yes"})
autos["fuel_type"].map({"benzin":"petrol", "andere":"others","elektro":"electric","petrol":"petrol","others":"others","electric":"electric"})
autos["gearbox"].map({"manuell":"manual","automatik":"automatic","manual":"manual","automatic":"automatic"})
autos["vehicle_type"].map({"limousine":"limousine","kleinwagen":"small_car","kombi":"van","bus":"bus","small_car":"small_car","van":"van","cabrio":"convertible","convertible":"convertible","suv":"suv","coupe":"coupe","andere":"others","others":"others"})

0                bus
1          limousine
2          limousine
3          small_car
4                van
5                bus
6          limousine
7          limousine
8                NaN
9                bus
10               NaN
11               NaN
12         small_car
13         limousine
14         small_car
15             coupe
16               van
17             coupe
18               bus
19               suv
20               van
21       convertible
22       convertible
23               bus
24         limousine
25               van
26         small_car
28         small_car
29               van
30         small_car
            ...     
49968      limousine
49969            suv
49970            bus
49971      small_car
49972            bus
49973    convertible
49975      small_car
49976            van
49977      limousine
49978      limousine
49979      small_car
49980            van
49981            van
49982      small_car
49983      small_car
49985            NaN
49986      li

### Date Columns

There are a number of values in this dataset dealing with dates
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

In [20]:
autos[['date_crawled','ad_created','last_seen','registration_month','registration_year']][0:5]

date_crawled           ad_created            last_seen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   

   registration_month  registration_year  
0                   3               2004  
1                   6               1997  
2                   7               2009  
3                   6               2007  
4                   7               2003

In the first three columns, the date is shown in a yr-mth-dy format and this is constantly the first 10 characters of each value in those columns. We can use this to explore these columns

In [21]:
autos["date_crawled"].str[:10].value_counts(dropna=False).sort_index()

2016-03-05    1230
2016-03-06     682
2016-03-07    1749
2016-03-08    1617
2016-03-09    1607
2016-03-10    1563
2016-03-11    1582
2016-03-12    1793
2016-03-13     761
2016-03-14    1775
2016-03-15    1665
2016-03-16    1438
2016-03-17    1536
2016-03-18     627
2016-03-19    1689
2016-03-20    1840
2016-03-21    1815
2016-03-22    1602
2016-03-23    1565
2016-03-24    1425
2016-03-25    1535
2016-03-26    1564
2016-03-27    1510
2016-03-28    1693
2016-03-29    1656
2016-03-30    1636
2016-03-31    1546
2016-04-01    1636
2016-04-02    1723
2016-04-03    1875
2016-04-04    1772
2016-04-05     636
2016-04-06     154
2016-04-07      68
Name: date_crawled, dtype: int64

This column ranges over a month with values from 5th of March, 2016 to 7th of April, 2016. With the 3rd of April having the highest frequency and the 7th of April having the lowest. Values at the later dates(last three values) tended to a lower end. A look at the last_seen column might hold more information

In [22]:
autos["last_seen"].str[:10].value_counts(dropna=False).sort_index()

2016-03-05       52
2016-03-06      210
2016-03-07      262
2016-03-08      360
2016-03-09      466
2016-03-10      518
2016-03-11      601
2016-03-12     1155
2016-03-13      432
2016-03-14      612
2016-03-15      771
2016-03-16      799
2016-03-17     1364
2016-03-18      357
2016-03-19      769
2016-03-20     1003
2016-03-21     1002
2016-03-22     1038
2016-03-23      900
2016-03-24      960
2016-03-25      933
2016-03-26      816
2016-03-27      760
2016-03-28     1013
2016-03-29     1085
2016-03-30     1203
2016-03-31     1155
2016-04-01     1107
2016-04-02     1210
2016-04-03     1224
2016-04-04     1189
2016-04-05     6059
2016-04-06    10772
2016-04-07     6408
Name: last_seen, dtype: int64

This column shows a higher set of values in the last three days. As this behaviour was also seen in the previous column, we can infer that the problem is from their common factor, they are both crawler values.

In [23]:
autos["ad_created"].str[:10].value_counts(dropna=False).sort_index()

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
2015-12-30       1
2016-01-03       1
2016-01-07       1
2016-01-10       2
2016-01-13       1
2016-01-14       1
2016-01-16       1
2016-01-22       1
2016-01-27       3
2016-01-29       1
2016-02-01       1
2016-02-02       2
2016-02-05       2
2016-02-07       1
2016-02-08       1
2016-02-09       1
2016-02-11       1
2016-02-12       2
2016-02-14       2
2016-02-16       1
2016-02-17       1
2016-02-18       2
2016-02-19       3
2016-02-20       2
2016-02-21       3
              ... 
2016-03-09    1610
2016-03-10    1549
2016-03-11    1598
2016-03-12    1785
2016-03-13     826
2016-03-14    1709
2016-03-15    1652
2016-03-16    1463
2016-03-17    1519
2016-03-18     660
2016-03-19    1636
2016-03-20    1843
2016-03-21    1825
2016-03-22    1593
2016-03-23    1557
2016-03-24    1422
2016-03-25    1542
2016-03-26    1567
2016-03-27    1505
2016-03-28    1699
2016-03-29    1653
2016-03-30  

There are unique entries to this column, showing the number of ads created during these dates. The values started slowly from a single ad on the 11th of June, 2015 till increased traction from the 5th of March, 2016. The latest date on this column is the 7th of April, 2016.

***********************************************
Rather than using .str[] we can extract the inormation we want to its own column
***********************************************

In [24]:
date_created = autos["ad_created"].str[:10].str.replace("-","").astype(int)
date_ended = autos["last_seen"].str[:10].str.replace("-","").astype(int)
date_extracted = autos["date_crawled"].str[:10].str.replace("-","").astype(int)

In [25]:
autos["date_created"] = date_created
autos["date_ended"] = date_ended
autos["date_extracted"] = date_extracted

In [26]:
autos.iloc[:,-3:].head(10)

date_created  date_ended  date_extracted
0      20160326    20160406        20160326
1      20160404    20160406        20160404
2      20160326    20160406        20160326
3      20160312    20160315        20160312
4      20160401    20160401        20160401
5      20160321    20160406        20160321
6      20160320    20160323        20160320
7      20160316    20160407        20160316
8      20160322    20160326        20160322
9      20160316    20160406        20160316

This format is more suitable for other analysis cases where manipulation or the difference between dates is needed

In [27]:
autos["registration_year"].value_counts(dropna=False).sort_index()

1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

This column contains a number of invalid entries. The column is for the The year in which the car was first registered. This year should ideally not be more than the date the crawler ended (2016). Also there are a number of entries showing years earlier than the 1900's.

Being conservative in our cleaning, we can limit our valid entries to a range of 1900 to 2016.

In [28]:
autos["registration_month"].value_counts(dropna=False).sort_index()

0     4480
1     3219
2     2937
3     5003
4     4036
5     4031
6     4271
7     3857
8     3126
9     3330
10    3588
11    3313
12    3374
Name: registration_month, dtype: int64

Here we see 13 unique values. As this column is to denote the months of the year we can remove as it is not valid for our dataset analysis.

## Cleaning

In [29]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [30]:
autos = autos[autos["registration_month"].between(1,12)]

In [31]:
autos["registration_year"].describe()

count    42655.000000
mean      2003.195921
std          6.939793
min       1927.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [32]:
autos["registration_month"].describe()

count    42655.000000
mean         6.377119
std          3.351886
min          1.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

## Analyzing Brands

In [33]:
autos["brand"].value_counts(dropna=False)

volkswagen        8882
bmw               4778
opel              4432
mercedes_benz     4254
audi              3755
ford              2932
renault           1982
peugeot           1280
fiat              1085
seat               783
skoda              727
nissan             660
mazda              637
smart              616
citroen            589
toyota             567
hyundai            437
mini               399
sonstige_autos     398
volvo              396
mitsubishi         342
honda              334
kia                319
alfa_romeo         282
porsche            273
suzuki             256
chevrolet          245
chrysler           148
dacia              118
jeep               101
daihatsu           100
land_rover          93
subaru              86
saab                73
jaguar              67
daewoo              64
rover               56
lancia              43
trabant             43
lada                23
Name: brand, dtype: int64

In [34]:
autos["brand"].value_counts(dropna=False).head(10)

volkswagen       8882
bmw              4778
opel             4432
mercedes_benz    4254
audi             3755
ford             2932
renault          1982
peugeot          1280
fiat             1085
seat              783
Name: brand, dtype: int64

Looking at the brands column, we see 40 unique values listed. We will look at the 10 of the most common brands for further brands

In [35]:
common_brands = autos["brand"].value_counts(dropna=False).head(10).index

#### Analyzing for mean price

In [36]:
brand_mean_prices = {}

for row in common_brands:
    brand = autos[autos["brand"] == row]
    mean_price = brand["price"].mean()
    brand_mean_prices[row] = int(mean_price)    
    
sorted(brand_mean_prices.items(), key=lambda x: x[1], reverse=True)

[('audi', 9776),
 ('mercedes_benz', 8882),
 ('bmw', 8652),
 ('volkswagen', 5752),
 ('seat', 4568),
 ('ford', 3957),
 ('peugeot', 3221),
 ('opel', 3167),
 ('fiat', 2950),
 ('renault', 2616)]

The Audi, Mercedes Benz, BMW brands are the most expensive on average. Volkswagen which is the most common entry in this dataset comes as the 4th most expensive. Seat which was the 10th most common is the 5th most expensive

#### Analyzing for mileage

In [37]:
brand_mean_mileage = {}

for row in common_brands:
    brand = autos[autos["brand"] == row]
    mean_mileage = brand["odometer_km"].mean()
    brand_mean_mileage[row] = int(mean_mileage)    
    
sorted(brand_mean_mileage.items(), key=lambda x: x[1], reverse=True)

[('bmw', 132489),
 ('mercedes_benz', 130927),
 ('opel', 128752),
 ('audi', 128593),
 ('volkswagen', 128183),
 ('renault', 127820),
 ('peugeot', 126269),
 ('ford', 124009),
 ('seat', 120791),
 ('fiat', 116188)]

We can see some similarities with the previous mean price values, as BMW and Mercedes Benz top the table for the cars with the most mileage on average showing that the cars have some extra even when aged. Compiling the data in a tabular form could help with this analysis

In [38]:
bmp_series = pd.Series(brand_mean_prices).sort_values(ascending=False)
bmm_series = pd.Series(brand_mean_mileage).sort_values(ascending=False)

In [39]:
mean_data = pd.DataFrame(bmm_series, columns = ["mean_mileage"])

In [40]:
mean_data["mean_price"] = bmp_series
mean_data

mean_mileage  mean_price
bmw                  132489        8652
mercedes_benz        130927        8882
opel                 128752        3167
audi                 128593        9776
volkswagen           128183        5752
renault              127820        2616
peugeot              126269        3221
ford                 124009        3957
seat                 120791        4568
fiat                 116188        2950

From this table, we see a trend for brands with higher price having a higher mileage as people keep them longer before resale. The opel brand is a notable exception in the top 5.

### Analyzing brand-model combinations

In [63]:
autos.groupby(["brand", "model"]).size().sort_values(ascending=False).head()

brand       model 
volkswagen  golf      3368
bmw         3er       2427
volkswagen  polo      1410
opel        corsa     1403
volkswagen  passat    1245
dtype: int64

This shows the five most common cars put up on the website. Volkswagen again showing its value as it holds three spots on the list. 